### PR TITLE
Support aud claim verification for arrays

### PR DESCRIPTION
**Pull request by kekoav**
This references issue #184 .

The goal is to support compliance with [RFC 7519 Section 4.1](https://tools.ietf.org/html/rfc7519#section-4.1). 

JWT audience claims `aud` should support an array of valid `aud` values in the token. When validating a token, a single `aud` value is provided and compared against the array of aud claims.

Care was taken to maintain backward compatibility with the `string` version of `aud` as it works today. This is also part of the spec, so this should not be a breaking change at all. A single string is supported and is treated as a list of one item.

The library already supports creation of JWTs with this structure using the `MapClaims`, but obviously the `StandardClaims` struct is too rigid for this use case.

Please give me your feedback- I'm happy to improve this pull request as needed.

### CODE REVIEW

**Comment by kekoav** [comment]
So the build fail is on Go 1.3 it's due to a poor implementation of the `ConstantTimeCompare()` function which assumed equal lengths, and panics 😮 if they are different . This fail should occur on any comparison using this method where the values have dissimilar lengths when using Go 1.3. I could fix this by performing the length check before calling `ConstantTimeCompare()`. This is a bit annoying because it adds an extra check for the majority of users Go 1.4+, but if we need to retain Go 1.3 support then we . 

## Go 1.3 [pkg/crypto/subtle/constant_time.go](https://github.com/golang/go/blob/release-branch.go1.3/src/pkg/crypto/subtle/constant_time.go#L9-L24)
```go
// ConstantTimeCompare returns 1 iff the two equal length slices, x
// and y, have equal contents. The time taken is a function of the length of
// the slices and is independent of the contents.
func ConstantTimeCompare(x, y []byte) int {
	if len(x) != len(y) {
		panic("subtle: slices have different lengths")
	}
...
```

## Fixed in Go 1.4 [pkg/crypto/subtle/constant_time.go](https://github.com/golang/go/blob/release-branch.go1.4/src/crypto/subtle/constant_time.go#L9-L24)

```go
// ConstantTimeCompare returns 1 iff the two slices, x
// and y, have equal contents. The time taken is a function of the length of
// the slices and is independent of the contents.
func ConstantTimeCompare(x, y []byte) int {
	if len(x) != len(y) {
		return 0
	}
...
```

It would be nice to have a more robust suite of tests around claim validation - this issue should cause panics in 1.3 if a JWT claim for `aud`, or `iss` is attempted to be validated against strings with a different length.

**Comment by dgrijalva** [comment]
Might make sense to figure out sunsetting 1.3 support.  How do you think that interacts with semver?

File: map_claims.go [hunk 1]
```diff
@@ -3,18 +3,44 @@ package jwt
...
+// If the aud claim is a string, this method will return true if the value matches exactly.
+// If the aud claim is a slice of strings, this method will return true if the value exactly matches any of the items in the slice.
 // If required is false, this method will return true if the value matches or is unset
 func (m MapClaims) VerifyAudience(cmp string, req bool) bool {
-	aud, _ := m["aud"].(string)
-	return verifyAud(aud, cmp, req)
+	aud := m["aud"]
+
+	switch aud.(type) {
+	default:
```
Comment (kekoav) [inline map_claims.go]: It works first, I saw this first at https://golang.org/doc/effective_go.html#type_switch . I agree it probably reads better last though.

**Comment by kekoav** [comment]
@dgrijalva 
> Might make sense to figure out sunsetting 1.3 support. How do you think that interacts with semver?

I'm not sure exactly about what to do with Semver for dropping a supported language version. I think it just needs to be communicated clearly. My initial thought is since there are no breaking API changes, this is definitely not a major release. Dropping Go 1.3 is notable enough that you may want it to be a minor point release (3.1).

**Comment by kekoav** [comment]
@dgrijalva any feedback on this PR? I'm happy to improve this PR based on your feedback.

File: claims.go [hunk 2]
```diff
@@ -90,15 +90,19 @@ func (c *StandardClaims) VerifyNotBefore(cmp int64, req bool) bool {
...
 	}
-	if subtle.ConstantTimeCompare([]byte(aud), []byte(cmp)) != 0 {
-		return true
-	} else {
-		return false
+
+	matchesAny := false
+	for _, a := range aud {
+		if subtle.ConstantTimeCompare([]byte(a), []byte(cmp)) != 0 {
+			matchesAny = true
```
Comment (tuxlife) [inline claims.go]: Why not `return true` and at the end of this function `return false`?

Comment (leonklingele) [inline claims.go]: You can break here, or simply `return true`

File: claims.go [hunk 3]
```diff
@@ -90,15 +90,19 @@ func (c *StandardClaims) VerifyNotBefore(cmp int64, req bool) bool {
 
 // ----- helpers
 
-func verifyAud(aud string, cmp string, required bool) bool {
-	if aud == "" {
+func verifyAud(aud []string, cmp string, required bool) bool {
+	if len(aud) <= 0 || aud[0] == "" {
```
Comment (leonklingele) [inline claims.go]: Why `<=`?

**Comment by dgrijalva** [comment]
A modified version of this has landed in the v4 branch.

**Pull request closed without merging**